### DISCUSSION TITLE
Coding advice - Mortgage calculator

### DISCUSSION THREAD

**Discussion by ozdotcom** [CarlosPietrasanta/Tafe230130#4] [Q&A] [answered]
Hi Jason and Carlos,

Any chance either of you have time to look over my code and give some feedback.

Would really appreciate a few suggestions on improving my code and calculating the repayments.

	public MortgageCalculator()
		{
			this.InitializeComponent();
		}
		private async void CalculateButton_Click(object sender, RoutedEventArgs e)
		{
			int years, months;
			double principle, yearRate, monthRate;
			double mRepayment;
			try
			{
				years = int.Parse(yearsTextBox.Text);
			}
			catch (Exception ex)
			{
				MessageDialog dialogMessage = new MessageDialog("Error please enter the years" + ex.Message);
				await dialogMessage.ShowAsync();
				yearsTextBox.Focus(FocusState.Programmatic);
				yearsTextBox.SelectAll();
				return;
			}
			try
			{
				months = int.Parse(monthsTextBox.Text);
			}
			catch (Exception ex)
			{
				MessageDialog dialogMessage = new MessageDialog("Error please enter the months" + ex.Message);
				await dialogMessage.ShowAsync();
				monthsTextBox.Focus(FocusState.Programmatic);
				return;
			}
			try
			{
				principle = double.Parse(principleTextBox.Text);
			}
			catch (Exception ex)
			{
				MessageDialog dialogMessage = new MessageDialog("Error please enter a principle loan amount" + ex.Message);
				await dialogMessage.ShowAsync();
				principleTextBox.Focus(FocusState.Programmatic);
				return;
			}
			try
			{
				yearRate = double.Parse(yearlyInterestRateTextBox.Text);
			}
			catch (Exception ex)
			{
				MessageDialog dialogMessage = new MessageDialog("Error please enter the yearly interest rate" + ex.Message);
				await dialogMessage.ShowAsync();
				yearlyInterestRateTextBox.Focus(FocusState.Programmatic);
				return;
			}
			try
			{
				monthRate = double.Parse(monthlyInterestTextBox.Text);
			}
			catch (Exception ex)
			{
				MessageDialog dialogMessage = new MessageDialog("Error please enter the yearly interest rate" + ex.Message);
				await dialogMessage.ShowAsync();
				monthlyInterestTextBox.Focus(FocusState.Programmatic);
				return;
			}
			//mRepayment = calcMonthlyPayment();
			principle = double.Parse(principleTextBox.Text);
			yearRate = double.Parse(yearlyInterestRateTextBox.Text);
			monthRate = yearRate / 12;
			months = years * 12; 
			years = int.Parse(yearsTextBox.Text);

			// mRepayment =  principle * monthRate / (1 -Math.Pow(1 + monthRate, -months));
			// mRepaymentTextBox.Text = mRepayment;
			// monthlyInterestTextBox.Text = mRepayment.ToString("C");
			// monthRate = double.Parse(monthlyInterestTextBox.Text);
			// mRepaymentTextBox = mRepayment;
		}

		private void calcMonthlyPayment(int principle, int years, double monthRate)
		{
			//mRepaymentTextBox = principle[monthRate(1 + monthRate) ^ months - 1];
		}
	}

Thanks so much,

Sharryn
@Jasonlovedataanalyst 
@CarlosPietrasanta

**Comment by ozdotcom** [comment]
I have updated the calculate to this.
	----------------------------------------last try catch
try
			{
				monthRate = double.Parse(monthlyInterestTextBox.Text);
			}
			catch (Exception ex)
			{
				MessageDialog dialogMessage = new MessageDialog("Error please enter the yearly interest rate" + ex.Message);
				await dialogMessage.ShowAsync();
				monthlyInterestTextBox.Focus(FocusState.Programmatic);
				return;
			}
// Calculate monthly payment...(added this to the CalculateButton_Click event)
			principle = double.Parse(principleTextBox.Text);
			yearRate = double.Parse(yearlyInterestRateTextBox.Text);
			monthRate = yearRate / 12;
			months = years * 12;
			double mRepayment = calcMonthlyPayment(principle, years, monthRate);
			mRepaymentTextBox.Text = mRepayment.ToString("C");
		}
// updated the calcMonthlyPayment method-----------------------

		private double calcMonthlyPayment(double principle, int years, double monthRate)
		{
			int months = years * 12;
			double mRepayment = principle * (monthRate * Math.Pow(1 + monthRate, months)) / (Math.Pow(1 + monthRate, months) - 1);
			//mRepaymentTextBox.Text = mRepayment.ToString("C");
			return mRepayment;
		}

It appears to be working ok at this stage.

I will add the exit button code.

**Comment by CarlosPietrasanta** [comment, marked as answer]
Hi Sharryn, your code is looking good!

Just 2 little things:

- What about using "monthlyRepayment" instead of "mRepayment"? I think it would make it a bit easier to understand at first sight. 

- In the try-catch block where the code is parsing a double from monthRate.TextBox, the Error Dialog says "Please enter the _yearly_ interest rate". Is that a copy-paste that you forgot to modify? Or is something else going on that I couldn't get?

Besides that, seems to be alright!

**Reply by ozdotcom** [reply]
Thanks Carlos 
I will update your suggestion monthlyRepayment. 

And good pick up on the error message, thanks. 

Sharryn

**Reply by Jasonlovedataanalyst** [reply]
Hell Sharryn, your code looks great , would you mind to share your your xaml code , I want to keep the size to  be consistent with yours.

I am currently just simply using the stack panel to be nested with each other

**Reply by ozdotcom** [reply]
Hi Jason 

Thanks, 
I have a few issues with the calculation still but will get there.

Here is the xaml code for you to keep with the theme.
Carlos shared the lighter color, as we were having trouble with the button disappearing in the background when hovered over.
	
`<Page.Resources>
		<Style TargetType="TextBox">
			<Setter Property="Background" Value="LightCyan" />
			<Setter Property="Foreground" Value="DarkBlue" />
			<Setter Property="Margin" Value="10" />
			<Setter Property="MinWidth" Value="200" />
			<Setter Property="BorderBrush" Value="DarkBlue"/>
		</Style>
		<Style TargetType="TextBlock">
			<Setter Property="Foreground" Value="DarkBlue" />
			<Setter Property="Margin" Value="10" />
			<Setter Property="MinWidth" Value="100" />
		</Style>
		<Style TargetType="Button">
			<Setter Property="Background" Value="LightCyan" />
			<Setter Property="Foreground" Value="DarkBlue" />
			<Setter Property="Padding" Value="10" />
			<Setter Property="Margin" Value="20" />
			<Setter Property="MinWidth" Value="120" />
			<Setter Property="BorderBrush" Value="Black"/>
		</Style>
	</Page.Resources>`

`<Grid Background="DarkCyan">
		<StackPanel >
            
            <TextBlock x:Name="Title" Text="Mortgage Calculator" FontFamily="Monotype Corsiva" FontSize="72" FontWeight="Bold" HorizontalAlignment="Center">
            </TextBlock>

            <StackPanel Orientation="Horizontal" Width="451" >
                <TextBlock x:Name="Principle" Text="Principal:" FontSize="22" FontWeight="Bold" HorizontalAlignment="Right"/>
                <TextBox x:Name="principleTextBox" PlaceholderText="100000" />
            </StackPanel>

            <StackPanel Orientation="Horizontal" Width="448">
                <TextBlock  Text="Years:" FontSize="22" FontWeight="Bold"  />
                <TextBox x:Name="yearsTextBox" Text="30"  />
            </StackPanel>

            <StackPanel Orientation="Horizontal" Width="517">
                <TextBlock x:Name="Months"  Text="And Months:" FontSize="22" FontWeight="Bold" HorizontalAlignment="Right" />
                <TextBox x:Name="monthsTextBox" PlaceholderText="0" />
            </StackPanel>

            <StackPanel Orientation="Horizontal" Width="659">
                <TextBlock x:Name="interestRateTextBlock" Text="Yearly Interest Rate:"  FontSize="22" FontWeight="Bold" HorizontalAlignment="Right"/>
                <TextBox x:Name="yearlyInterestRateTextBox"  PlaceholderText="3.5" />
            </StackPanel>

            <StackPanel Orientation="Horizontal" Width="709">
                <TextBlock x:Name="MRateTextBlock"  Text="Monthly Interest Rate:"  FontSize="22" FontWeight="Bold" HorizontalAlignment="Right"/>
                <TextBox x:Name="monthlyInterestTextBox" PlaceholderText="0.0029%"  />
            </StackPanel >

            <StackPanel Orientation="Horizontal" Width="718">
                <TextBlock x:Name="MRepaymentTextBlock" Text="Monthly Repayment $:" FontSize="22" FontWeight="Bold" HorizontalAlignment="Right"/>
                <TextBox x:Name="mRepaymentTextBox" PlaceholderText="4,490.45" />
            </StackPanel>

            <StackPanel Orientation="Horizontal" Width="306">
                <Button x:Name="calculateButton" Content="Calculate" Click="CalculateButton_Click"  />
                <Button x:Name="exitButton" Content="Exit" Click="exitButton_Click" />
            </StackPanel>

        </StackPanel>
    </Grid>`

Sharryn
